validation: flush undo files after last block write

kallewoof opened the pull request:
Fixes #17890. Replaces #17892.

Data files (`{blk|rev}<number>.dat`) pre-allocate space as they are written, and then trims down to the final size once they move on to the next sequence ("finalized flush"). The code currently assumes (incorrectly) that blk and rev files finish at the same time, but because blk files are written as blocks come in, and rev files are written in block height order, rev files end up being written to for awhile after moving on to the next block file, resulting in pre-allocation and waste of up to 1 MB of space per rev file.

The exact point at which rev file writing finishes is the highest height block found inside the corresponding block file, which is already available in the CBlockFileInfo vector. This PR moves finalized flushing of undo files to to directly after the undo data for the previous block file has been written.

There is a branch with annotation that demonstrates how this is handling flushing here: https://github.com/kallewoof/bitcoin/tree/200124-rev-files-annotated

kallewoof committed:
validation: delay flushing undo files in syncing node case

Data files are pre-allocated, and upon flush/finalization, they are trimmed down to their resulting size.
Block (blk) files are written to disk as blocks come in, which is often out of order, whereas undo (rev) files are written sequentially, as each block is added to the top of the chain.
When a block file hits the size limit, the system flushes and trims the file down to its final size, and moves on to the next block file.

Case 1: blocks are added to the chain as they come in (synced up node case) -- in this case, we will flush and finalize the undo file together with the block file.

Case 2: blocks are added to the chain after they have been downloaded (syncing node case) -- in this case, we postpone finalizing the undo file until we know the undo data for the last block in the file has been written to disk.

Sjors commented on src/validation.cpp:
```diff
@@ -1757,6 +1761,12 @@ static bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationSt
         if (!UndoWriteToDisk(blockundo, _pos, pindex->pprev->GetBlockHash(), chainparams.MessageStart()))
             return AbortNode(state, "Failed to write undo data");
+        // rev files are written in block height order, whereas blk files are written as blocks come in (often out of order)
+        // we want to flush the rev (undo) file once we've written the last block, which is indicated by the last height
```
The last height, or the maximum height?

kallewoof replied:
They are the same:

https://github.com/bitcoin/bitcoin/blob/28fbe68fdcac2a06f359b1e48555a3d23015c2b7/src/chain.h#L85-L86

Sjors reviewed (commented):
Concept ACK. This seems more elegant than #17892, even though it involves more code.

kallewoof commented:
Synced two datadirs (master and this branch) to block 484516, which resulted in 995 rev files, master using 38277312 bytes, and this branch 35573680 bytes. The average wasted space per rev file comes out at 2717 bytes (~3kb), which is quite a bit lower than my assumptions (if it was completely random, it should be ~500kb per rev file).

Copying blocks/rev* into a separate dir reveals a few kb lower for this branch -- 35570968 -- caused by the latest rev file having some pre-allocation. The same for the master datadir results in 36547408 bytes.

Not sure why it's not higher, but still. I don't think a backport is even necessary, but fixing this bug and adding a one time sweep/finalize of all rev files is something that we should do.

fjahr commented:
utACK 48081229c10969ac5993c826c8bd7687cd0503e5

Agree, looks cleaner than #17892!

vasild commented:
I looked at the code and just want to confirm that I got it right. Is the following example correct?

Let's assume for simplicity that `blk*.dat` and `rev*.dat` files can contain up to 3 blocks.
Initially just the genesis block is in `blk01` and `rev01`. Then:
* block at height 3 is downloaded
  * it is written to `blk01`
  * undo for it will be written to `rev01`, but not yet
* block at height 4 is downloaded
  * it is written to `blk01`
  * undo for it will be written to `rev01`, but not yet
* block at height 5 is downloaded
  * `blk01` is finalized as it is full with 3 blocks
  * the downloaded block with height 5 is written to `blk02`
  * its undo will eventually be in `rev02`
* block at height 2 is downloaded
  * it is written to `blk02`
  * its undo is written to `rev02`
  * now that the block at height 3 is chained, undo for it is written to `rev01`
  * now that the block at height 4 is chained, undo for it is written to `rev01`
  * now that the block at height 5 is chained, undo for it is written to `rev02`

vasild commented on src/validation.cpp:
```diff
@@ -1731,19 +1731,23 @@ DisconnectResult CChainState::DisconnectBlock(const CBlock& block, const CBlockI
         AbortNode("Flushing block file to disk failed. This is likely the result of an I/O error.");
     }
+    // the undo file is actually finalized at a later stage, but we do want to flush it along with the block file
+    if (!fFinalize) FlushUndoFile(nLastBlockFile);
```
We want to flush the rev file along with the block file, but we are not going to do that if `fFinalize` is `true`. What about calling `FlushUndoFile(nLastBlockFile)` unconditionally here?

kallewoof replied:
There's basically no reason to do a non-final flush of the undo file, except on shutdown (which happens now). This is the same for the block file (since the case you are arguing is for a finalized flush).

vasild replied:
Ok, then this turns into a nit, feel free to ignore: maybe update the comment to be in line with the code: "we do want to flush it along with the block file _in non-finalizing flushes_".

kallewoof replied:
I didn't see this comment until now. I'll update the code in the next push.

vasild commented on src/validation.cpp:
```diff
@@ -1757,6 +1761,12 @@ static bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationSt
+        // we want to flush the rev (undo) file once we've written the last block, which is indicated by the last height
+        // in the block file info as below
+        if (_pos.nFile < nLastBlockFile && static_cast<uint32_t>(pindex->nHeight) == vinfoBlockFile[_pos.nFile].nHeightLast) {
+            FlushUndoFile(_pos.nFile, true);
```
Maybe I am missing something, but it looks to me that if blocks are being downloaded in-order, then the `_pos.nFile < nLastBlockFile` condition will never be true because those will always be equal?

kallewoof replied:
I am pretty sure I tested this but I can't convince myself that you're wrong by looking at the code so I will test this again.

vasild replied:
The way I see it is that `_pos.nFile < nLastBlockFile && ` can (should) be dropped and the second condition suffices - if we are writing the last (highest) block in the rev file, then also finalize it because we are not going to touch it again.

kallewoof replied:
Sorry for tardiness, work kept me busy, but I managed to verify that this is *not* working as it should. I will fix soon.

vasild replied:
No worries. I started at this problem long enough and finally it cracked.

Indeed `_pos.nFile < nLastBlockFile` is never true if blocks come in-height-order and this needs to be adjusted. However, just removing that condition, as I suggest above, is not ok because then the second condition will be true for every in-height-order write and this will cause excessive flushing and finalizing. See the proposed [patch](https://github.com/bitcoin/bitcoin/pull/17994#issuecomment-599031098) for a fully working solution (I hope).

kallewoof replied:
I've got something but testing takes time (I don't wanna have to wait days each time so I'm juggling 300 GB full node dirs around).

vasild commented:
I reproduced the bug in a controlled, deterministic environment with a [unit test + tweaks](https://gist.github.com/vasild/8c06b3dbc493522f683a671d71b4c122). The current patch in this PR (4808122) indeed fixes it, but it introduces another problem - it would wrongly skip finalizing of the undo files in case the blocks are downloaded in-height-order.

There are two cases when we should finalize an undo file (IFF):
* we have just moved to a subsequent one (like with block files) - "normal" in-height-order scenario or
* we are coming back to a lower undo file from a higher undo file to complete an out-of-order block's undo and this undo slot is the last/final in the lower undo file.

The following patch (on top of master @ 9cc7eba1b) fixes it:

```diff
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -131,13 +131,39 @@ CTxMemPool mempool(&feeEstimator);
 // Internal stuff
 namespace {
     CBlockIndex* pindexBestInvalid = nullptr;
 
     RecursiveMutex cs_LastBlockFile;
     std::vector<CBlockFileInfo> vinfoBlockFile;
+
+    /** Last blk*.dat file that has been written to.
+     * Used to detect when we move from blkN to blkN+1 so we can trigger a
+     * finalization of blkN.
+     * Blocks are appended to blk files in the order they are received from
+     * the network (which may not be height order). When a blk file is about
+     * to become larger than MAX_BLOCKFILE_SIZE then we move to the next blk
+     * file and finalize the one we just left.
+     */
     int nLastBlockFile = 0;
+
+    /** Highest rev*.dat file that has been written to.
+     * Used to detect the last write to revN so it can be finalized.
+     * Undo files don't have a max size limit. For every block that resides in
+     * blkN, its corresponding undo resides in revN.
+     * However, the undo for a given block is only appended to the rev file
+     * after all blocks at a lower height have been stored together with their
+     * undos. It follows that block and undo data may be in different order
+     * in the corresponding blk and rev files, for example:
+     * blkN: block5, block7, block6
+     * revN: block5, block6, block7
+     * if block7 was received before block6.
+     * Also we may write to revN after writing to revM, for N < M which never
+     * happens with blk files.
+     */
+    int nHighestUndoFile = 0;
+
     /** Global flag to indicate we should check to see if there are
      *  block/undo files that should be deleted.  Set on startup
      *  or if we allocate more file space when we're in prune mode
      */
     bool fCheckForPruning = false;
 
@@ -1733,23 +1759,25 @@ DisconnectResult CChainState::DisconnectBlock(const CBlock& block, const CBlockI
     // move best block pointer to prevout block
     view.SetBestBlock(pindex->pprev->GetBlockHash());
 
     return fClean ? DISCONNECT_OK : DISCONNECT_UNCLEAN;
 }
 
-void static FlushBlockFile(bool fFinalize = false)
+static void FlushUndoFile(int undo_file, bool finalize = false)
 {
-    LOCK(cs_LastBlockFile);
-
-    FlatFilePos block_pos_old(nLastBlockFile, vinfoBlockFile[nLastBlockFile].nSize);
-    FlatFilePos undo_pos_old(nLastBlockFile, vinfoBlockFile[nLastBlockFile].nUndoSize);
+    FlatFilePos pos(undo_file, vinfoBlockFile[undo_file].nUndoSize);
+    if (!UndoFileSeq().Flush(pos, finalize)) {
+        AbortNode("Flushing undo file to disk failed. This is likely the result of an I/O error.");
+    }
+}
 
-    bool status = true;
-    status &= BlockFileSeq().Flush(block_pos_old, fFinalize);
-    status &= UndoFileSeq().Flush(undo_pos_old, fFinalize);
-    if (!status) {
+static void FlushBlockFile(bool fFinalize = false)
+{
+    LOCK(cs_LastBlockFile);
+    FlatFilePos pos(nLastBlockFile, vinfoBlockFile[nLastBlockFile].nSize);
+    if (!BlockFileSeq().Flush(pos, fFinalize)) {
         AbortNode("Flushing block file to disk failed. This is likely the result of an I/O error.");
     }
 }
 
 static bool FindUndoPos(BlockValidationState &state, int nFile, FlatFilePos &pos, unsigned int nAddSize);
 
@@ -1760,12 +1788,37 @@ static bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationSt
         FlatFilePos _pos;
         if (!FindUndoPos(state, pindex->nFile, _pos, ::GetSerializeSize(blockundo, CLIENT_VERSION) + 40))
             return error("ConnectBlock(): FindUndoPos failed");
         if (!UndoWriteToDisk(blockundo, _pos, pindex->pprev->GetBlockHash(), chainparams.MessageStart()))
             return AbortNode(state, "Failed to write undo data");
 
+        const unsigned thisHeight = static_cast<unsigned>(pindex->nHeight);
+
+        // Flush and finalize the preceding undo file in the case when blocks
+        // come in height order and we have moved to a subsequent undo file.
+        const bool movedToSubsequent = _pos.nFile > nHighestUndoFile;
+        const bool adjacentHeights =
+            thisHeight == vinfoBlockFile[nHighestUndoFile].nHeightLast + 1;
+
+        if (movedToSubsequent && adjacentHeights) {
+            FlushUndoFile(nHighestUndoFile, true);
+        }
+
+        // Flush and finalize the old undo file if we went back to write
+        // to it and we wrote its last entry.
+        const bool wroteToOldFile = _pos.nFile < nHighestUndoFile;
+        const bool isLastEntry = thisHeight == vinfoBlockFile[_pos.nFile].nHeightLast;
+
+        if (wroteToOldFile && isLastEntry) {
+            FlushUndoFile(_pos.nFile, true);
+        }
+
+        if (movedToSubsequent) {
+            nHighestUndoFile = _pos.nFile;
+        }
+
         // update nUndoPos in block index
         pindex->nUndoPos = _pos.nPos;
         pindex->nStatus |= BLOCK_HAVE_UNDO;
         setDirtyBlockIndex.insert(pindex);
     }
 
@@ -2291,12 +2344,13 @@ bool CChainState::FlushStateToDisk(
             }
             {
                 LOG_TIME_MILLIS("write block and undo data to disk", BCLog::BENCH);
 
                 // First make sure all block and undo data is flushed to disk.
                 FlushBlockFile();
+                FlushUndoFile(nHighestUndoFile);
             }
 
             // Then update all block file information (which may refer to block and undo files).
             {
                 LOG_TIME_MILLIS("write block index to disk", BCLog::BENCH);
 
```

Feel free to pick it in its entirety, just pieces of it, just the idea or propose another approach. I will test it.

[Annotated debug output with the bug present and with the bug fixed (by the above patch).](https://gist.github.com/vasild/34e342c869aff9a5a870b3ef3fff154c)

Run the unit test like `./src/test/test_bitcoin --run_test="*/download_blocks_out_of_order" --log_level=all -x 0 > /tmp/log`
Filter its output: `grep -w BBB /tmp/log |sed -E -e 's/^2020[^ ]+ //' -e 's/BBB //'`

I hope this helps!

kallewoof commented:
@vasild Tremendous work. Thank you very much. I will look closer at your patch, and also evaluate my solution over the next few days to see if it does what I want. (And compare it to yours.)

My solution was to patch this in two places; one for the case "rev file lags behind", and one for "rev file is up to date". This was very little code (note that the last commit is a NOMERGE to demonstrate the fix).

I see this, for a node that was a week or so behind, if I grep -v 'UpdateTip':
```
2020-03-19T12:34:31Z Pre-allocating up to position 0x7000000 in blk02004.dat
2020-03-19T12:34:32Z Pre-allocating up to position 0x8000000 in blk02004.dat
2020-03-19T12:34:33Z iterating nFile: last height = 622161; chain tip height = 621899 -> DO NOT FLUSH UNDO FILE
2020-03-19T12:34:33Z Leaving block file 2004: CBlockFileInfo(blocks=112, size=133860058, heights=621899...622161, time=2020-03-16...2020-03-19)
2020-03-19T12:34:34Z unflushed undo: [2003 2004]
2020-03-19T12:34:34Z Pre-allocating up to position 0x1000000 in blk02005.dat
2020-03-19T12:34:34Z New outbound peer connected: version: 70015, blocks=622190, peer=10 (full-relay)
2020-03-19T12:34:34Z Pre-allocating up to position 0x2000000 in blk02005.dat
2020-03-19T12:34:35Z Pre-allocating up to position 0x3000000 in blk02005.dat
2020-03-19T12:34:36Z Pre-allocating up to position 0x4000000 in blk02005.dat
2020-03-19T12:34:38Z Pre-allocating up to position 0x5000000 in blk02005.dat
2020-03-19T12:34:43Z Pre-allocating up to position 0x100000 in rev02005.dat
2020-03-19T12:34:45Z New outbound peer connected: version: 70015, blocks=622190, peer=11 (full-relay)
2020-03-19T12:34:46Z New outbound peer connected: version: 70015, blocks=622190, peer=12 (block-relay)
2020-03-19T12:34:46Z New outbound peer connected: version: 70015, blocks=622190, peer=13 (block-relay)
2020-03-19T12:34:48Z Pre-allocating up to position 0x6000000 in blk02005.dat
2020-03-19T12:34:51Z New outbound peer connected: version: 70015, blocks=622190, peer=14 (block-relay)
2020-03-19T12:35:09Z Pre-allocating up to position 0x7000000 in blk02005.dat
2020-03-19T12:35:10Z Synchronizing blockheaders, height: 622191 (~100.00%)
2020-03-19T12:37:49Z Synchronizing blockheaders, height: 622192 (~100.00%)
2020-03-19T12:39:53Z Pre-allocating up to position 0x200000 in rev02005.dat
2020-03-19T12:40:43Z Pre-allocating up to position 0x8000000 in blk02005.dat
2020-03-19T12:41:24Z Synchronizing blockheaders, height: 622193 (~100.00%)
2020-03-19T12:43:44Z Pre-allocating up to position 0x300000 in rev02005.dat
2020-03-19T12:44:01Z Pre-allocating up to position 0x400000 in rev02005.dat
2020-03-19T12:44:02Z Pre-allocating up to position 0x600000 in rev02003.dat
2020-03-19T12:44:03Z Pre-allocating up to position 0x700000 in rev02003.dat
2020-03-19T12:44:03Z Pre-allocating up to position 0x800000 in rev02003.dat
2020-03-19T12:44:04Z Pre-allocating up to position 0x900000 in rev02003.dat
2020-03-19T12:44:04Z Pre-allocating up to position 0xa00000 in rev02003.dat
2020-03-19T12:44:05Z Pre-allocating up to position 0xb00000 in rev02003.dat
2020-03-19T12:44:06Z Pre-allocating up to position 0x500000 in rev02005.dat
2020-03-19T12:44:06Z Pre-allocating up to position 0x200000 in rev02004.dat
2020-03-19T12:44:06Z Pre-allocating up to position 0xc00000 in rev02003.dat
2020-03-19T12:44:07Z Pre-allocating up to position 0xd00000 in rev02003.dat
2020-03-19T12:44:08Z Pre-allocating up to position 0xe00000 in rev02003.dat
2020-03-19T12:44:09Z Pre-allocating up to position 0x300000 in rev02004.dat
2020-03-19T12:44:09Z Pre-allocating up to position 0x600000 in rev02005.dat
2020-03-19T12:44:10Z Pre-allocating up to position 0xf00000 in rev02003.dat
2020-03-19T12:44:11Z Pre-allocating up to position 0x400000 in rev02004.dat
2020-03-19T12:44:12Z Pre-allocating up to position 0x500000 in rev02004.dat
2020-03-19T12:44:12Z Pre-allocating up to position 0x1000000 in rev02003.dat
2020-03-19T12:44:13Z Pre-allocating up to position 0x1100000 in rev02003.dat
2020-03-19T12:44:14Z *** FLUSH UNDO FILE 2003 ***
2020-03-19T12:44:14Z finalize undo 2003
2020-03-19T12:44:14Z unflushed undo: [2004 2005]
2020-03-19T12:44:14Z Pre-allocating up to position 0x600000 in rev02004.dat
2020-03-19T12:44:15Z Pre-allocating up to position 0x700000 in rev02004.dat
2020-03-19T12:44:16Z Pre-allocating up to position 0x800000 in rev02004.dat
2020-03-19T12:44:16Z Pre-allocating up to position 0x900000 in rev02004.dat
2020-03-19T12:44:16Z Leaving InitialBlockDownload (latching to false)
2020-03-19T12:44:18Z Pre-allocating up to position 0x700000 in rev02005.dat
2020-03-19T12:44:19Z Pre-allocating up to position 0xa00000 in rev02004.dat
2020-03-19T12:44:19Z Pre-allocating up to position 0xb00000 in rev02004.dat
2020-03-19T12:44:20Z Pre-allocating up to position 0xc00000 in rev02004.dat
2020-03-19T12:44:21Z Pre-allocating up to position 0xd00000 in rev02004.dat
2020-03-19T12:44:22Z Pre-allocating up to position 0xe00000 in rev02004.dat
2020-03-19T12:44:22Z Pre-allocating up to position 0x800000 in rev02005.dat
2020-03-19T12:44:23Z Pre-allocating up to position 0x900000 in rev02005.dat
2020-03-19T12:44:23Z Pre-allocating up to position 0xf00000 in rev02004.dat
2020-03-19T12:44:24Z Pre-allocating up to position 0x1000000 in rev02004.dat
2020-03-19T12:44:25Z Pre-allocating up to position 0xa00000 in rev02005.dat
2020-03-19T12:44:26Z Pre-allocating up to position 0x1100000 in rev02004.dat
2020-03-19T12:44:26Z Pre-allocating up to position 0xb00000 in rev02005.dat
2020-03-19T12:44:27Z *** FLUSH UNDO FILE 2004 ***
2020-03-19T12:44:27Z finalize undo 2004
2020-03-19T12:44:27Z unflushed undo: [2005]
2020-03-19T12:44:27Z Pre-allocating up to position 0xc00000 in rev02005.dat
2020-03-19T12:44:28Z Pre-allocating up to position 0xd00000 in rev02005.dat
2020-03-19T12:44:29Z Pre-allocating up to position 0xe00000 in rev02005.dat
2020-03-19T12:44:29Z Pre-allocating up to position 0xf00000 in rev02005.dat
2020-03-19T12:44:30Z Pre-allocating up to position 0x1000000 in rev02005.dat
2020-03-19T13:09:43Z Pre-allocating up to position 0x1100000 in rev02005.dat
2020-03-19T13:32:41Z unflushed undo: [2005]
2020-03-19T13:53:07Z iterating nFile: last height = 622198; chain tip height = 622198 -> FLUSH UNDO FILE
2020-03-19T13:53:07Z finalize undo 2005
2020-03-19T13:53:07Z unflushed undo: []
2020-03-19T13:53:07Z Leaving block file 2005: CBlockFileInfo(blocks=104, size=133278069, heights=621900...622198, time=2020-03-16...2020-03-19)
2020-03-19T13:53:07Z unflushed undo: []
2020-03-19T13:53:07Z Pre-allocating up to position 0x1000000 in blk02006.dat
2020-03-19T13:53:07Z Pre-allocating up to position 0x100000 in rev02006.dat
```

vasild commented on src/validation.cpp:
```diff
@@ -3222,6 +3253,16 @@ static bool FindBlockPos(FlatFilePos &pos, unsigned int nAddSize, unsigned int n
+                ChainActive().Tip()->nHeight,
+                vinfoBlockFile[nFile].nHeightLast == ChainActive().Tip()->nHeight ? "FLUSH UNDO FILE" : "DO NOT FLUSH UNDO FILE");
+            if (vinfoBlockFile[nFile].nHeightLast == ChainActive().Tip()->nHeight) {
+                FlushUndoFile(nFile, true);
```
This will flush the undo file before the corresponding block file which is flushed a [few lines below](https://github.com/bitcoin/bitcoin/pull/17994/files#diff-24efdb00bfbe56b140fb006b562cc70bR3279). Also visible in the log you pasted where `finalize undo 2005` appears before `Leaving block file 2005`.

I am not sure but I think we want to always flush the block file first and the undo file afterwards.

kallewoof replied:
I *think* the way it works now is fine; the "unflushed undo files" output is consistently showing the current opened one only, so it indicates the same.

Let me go through the steps:

* Sync state: HEIGHT = 100, BLK = 100, UNDO = 100, block file 0
* Block 101 comes in (we assume it does not fit into the current block file, so we will move to block file 1)
* First we want to store it to disk: 

https://github.com/bitcoin/bitcoin/blob/c4400e1b8a62269d04b57a46d68941fdbaae4303/src/validation.cpp#L3731-L3736

* We go find a position and end up in the next block file (1), at which point we flush the undo file (0):

https://github.com/bitcoin/bitcoin/blob/c4400e1b8a62269d04b57a46d68941fdbaae4303/src/validation.cpp#L3254-L3264

* Sync state: HEIGHT = 100, BLK = 101, UNDO = 100, block file 1
* We now want to accept block 101, and at this point we write the undo data (for block 101)

https://github.com/bitcoin/bitcoin/blob/c4400e1b8a62269d04b57a46d68941fdbaae4303/src/validation.cpp#L1773-L1787

* The condition does not trigger here, because we are synced up (as you pointed out), but we end up writing into undo file 1, and undo file 0 was finalized as it should.

Let me know if I got something wrong in the above (or if there are any other holes in the patch). I will update the log above once it moves to block file 2007.

vasild replied:
Yes, the above is correct (or at least coincides with my understanding for this elusive code!) - I think the code will do what you describe and that it is the correct behavior wrt finalizing the undo.

However, one piece is missing from the above - when do we flush block file 0?
> We go find a position and end up in the next block file (1), at which point we flush the undo file (0)

... and _after this_ we flush the blk file 0.

Just to be explicit, something like this may be warranted, delay the flush of the undo until after we have flushed the block file:

```diff
+    int thisUndoNeedsFlushingAndFinalizing = -1;
+
     if (!fKnown) {
         while (vinfoBlockFile[nFile].nSize + nAddSize >= MAX_BLOCKFILE_SIZE) {
             // when the undo file is keeping up with the block file, we flush it explicitly, here
             // when it is lagging behind (more blocks arrive than are being connected), we let the
             // undo block write case handle it
             LogPrintf("iterating nFile: last height = %d; chain tip height = %d -> %s\n",
                 vinfoBlockFile[nFile].nHeightLast,
                 ChainActive().Tip()->nHeight,
                 vinfoBlockFile[nFile].nHeightLast == ChainActive().Tip()->nHeight ? "FLUSH UNDO FILE" : "DO NOT FLUSH UNDO FILE");
             if (vinfoBlockFile[nFile].nHeightLast == ChainActive().Tip()->nHeight) {
-                FlushUndoFile(nFile, true);
+                thisUndoNeedsFlushingAndFinalizing = nFile;
             }
             nFile++;
             if (vinfoBlockFile.size() <= nFile) {
                 vinfoBlockFile.resize(nFile + 1);
             }
         }
         pos.nFile = nFile;
         pos.nPos = vinfoBlockFile[nFile].nSize;
     }
 
     if ((int)nFile != nLastBlockFile) {
         if (!fKnown) {
             LogPrintf("Leaving block file %i: %s\n", nLastBlockFile, vinfoBlockFile[nLastBlockFile].ToString());
         }
+        // Order is important: first FlushBlockFile(), then FlushUndoFile().
+        // FlushBlockFile() will flush internally nLastBlockFile and at this
+        // point I believe thisUndoNeedsFlushingAndFinalizing == nLastBlockFile
+        // so we will flush corresponding blk and undo files.
         FlushBlockFile(!fKnown);
+        if (thisUndoNeedsFlushingAndFinalizing != -1) {
+            FlushUndoFile(thisUndoNeedsFlushingAndFinalizing, true);
+        }
         nLastBlockFile = nFile;
     }
```

kallewoof replied:
Oh wait, I see what you mean -- you're concerned about the non-finalized undo flush happening after the finalized undo flush (called from within the finalized block flush), right?

I don't think this it's a problem to flush twice. The only difference between a finalized=true flush and a finalized=false flush is L87 in

https://github.com/bitcoin/bitcoin/blob/a3abe25f720104e4abf0d3f01f0b847929167b8b/src/flatfile.cpp#L81-L98

which means the non-finalized flush is a NOP, I think.

vasild replied:
No, sorry, I mentioned my concern (order of flushing of blk and undo files), but did not mention why is that a concern. Previously the code always flushed blkN before flushing revN (finalizing is not relevant for this):

https://github.com/bitcoin/bitcoin/blob/97b0687501cee77a9170f9e288755a5d268e9bd4/src/validation.cpp#L1747-L1748

In case of a crash (or power failure) between the two, we would end up with a block data on disk without a corresponding undo data. I presume this is ok in the current code.

Now, with the change in this PR we would flush revN first and blkN afterwards. This could potentially leave us with an undo data for some block(s) without the corresponding block data. There is at least this code which would get upset by that:

https://github.com/bitcoin/bitcoin/blob/97b0687501cee77a9170f9e288755a5d268e9bd4/src/validation.cpp#L4810

Now, I am not 100% sure if that is a real concern or not, but is relatively easy to tweak in the current patch so that it keeps flushing first blk and then undo (as before the patch).

vasild commented on src/validation.cpp:
```diff
@@ -3222,6 +3253,16 @@ static bool FindBlockPos(FlatFilePos &pos, unsigned int nAddSize, unsigned int n
 
     if (!fKnown) {
         while (vinfoBlockFile[nFile].nSize + nAddSize >= MAX_BLOCKFILE_SIZE) {
```
I find this `while` extremely confusing - it looks to me that it will never iterate more than once. We need a block size of more than 128MB (blk file size limit) to get this execute 2 or more times, so that one block spans multiple blk files (!?).

Maybe out of the scope of this PR, but I think we should `s/while/if/`. Is my understanding correct? If not then we will end up finalizing an empty undo file with this patch on the second and subsequent iterations.

kallewoof replied:
I dug into this. Apparently, the system is supposed to support multiple files *per block*. This was a change made 8 years ago in 5382bcf8cd23c36a435c29080770a79b5e28af42 by @sipa. It doesn't look like it is actually working, though, but I could be wrong.

sipa replied:
@kallewoof No, that commit adds support for multiple *blocks per file*, which is definitely used. There has never been support for multiple files per block.

kallewoof replied:
Oh, yeah you're right. But this while loop makes no sense to me still.

sipa replied:
I think it's just iterating to find the first block file with space for the new block.

kallewoof replied:
Okay, so assuming we one day decided to stop writing block files sequentially, we may have a situation where we try to `find block pos` for `nFile = 2007` where the last block file in existence is actually `nFile = 2010` or something `>2008`, in which case this while loop becomes necessary.

I don't really see that ever happening though, and it feels like someone got "block file writing order" (sequentially; we never write into older block files or into the middle of a file in between other stuff) and "connect order" (non-sequentially, as blocks are not stored in height order).

vasild replied:
Ok, for the purposes of this PR we just need to agree that this `while` loop will never iterate more than once (if it is possible that it iterates 2 or more times then I am wrong and this PR will have to be adjusted).

Observation: before the loop condition is evaluated for the first time `nFile` equals `nLastBlockFile`. Ie we start the search at the end.

So two things can happen:
1. We find space in the last file - the loop condition evaluates to `false` and the loop iterates 0 times
2. There is no space in the last file, the loop condition evaluates to `true` and the loop body "creates" a new empty block file (not yet on disk, but this is not relevant). The new empty block file has `nSize == 0` and so the next loop condition will for sure evaluate to `false` because `nSize + addedBlockSize` cannot be `>=` than 128 MiB.

> I think it's just iterating to find the first block file with space for the new block.

Yes, and given that it always starts the search at the last file it will never iterate more than once, as per above.

sipa replied:
I think in the original version of this code the loop was there so that after a restart a good file number would be found (even if it always reset to 0). I don't know if that's still needed.

vasild commented:
ACK ed34e00da. Code review and manual testing.

I tested this by applying the patch folded below which introduces a unit test and lowers the max size of blk files so that they can accommodate just 4 test blocks. It also adds some printouts which I manually analyzed after running the unit test as described at the bottom of https://github.com/bitcoin/bitcoin/pull/17994#issuecomment-599031098. I think this unit test does not qualify for an automated unit test.

Maybe squash all 4 commits in this PR to one.

<details>
<summary>unit test + tweaks on top of ed34e00da</summary>

```diff
--- i/src/flatfile.cpp
+++ w/src/flatfile.cpp
@@ -55,21 +55,23 @@ FILE* FlatFileSeq::Open(const FlatFilePos& pos, bool read_only)
 size_t FlatFileSeq::Allocate(const FlatFilePos& pos, size_t add_size, bool& out_of_space)
 {
     out_of_space = false;
 
     unsigned int n_old_chunks = (pos.nPos + m_chunk_size - 1) / m_chunk_size;
     unsigned int n_new_chunks = (pos.nPos + add_size + m_chunk_size - 1) / m_chunk_size;
+    LogPrintf("BBB FlatFileSeq::Allocate(): %s%05u@[%u..%zu]\n",
+            m_prefix, pos.nFile, pos.nPos, pos.nPos + add_size - 1);
     if (n_new_chunks > n_old_chunks) {
         size_t old_size = pos.nPos;
         size_t new_size = n_new_chunks * m_chunk_size;
         size_t inc_size = new_size - old_size;
 
         if (CheckDiskSpace(m_dir, inc_size)) {
             FILE *file = Open(pos);
             if (file) {
-                LogPrintf("Pre-allocating up to position 0x%x in %s%05u.dat\n", new_size, m_prefix, pos.nFile);
+                LogPrintf("BBB FlatFileSeq::Allocate(): Pre-allocating up to position %u in %s%05u\n", new_size, m_prefix, pos.nFile);
                 AllocateFileRange(file, pos.nPos, inc_size);
                 fclose(file);
                 return inc_size;
             }
         } else {
             out_of_space = true;
diff --git i/src/test/validation_block_tests.cpp w/src/test/validation_block_tests.cpp
index dae389a16..4f3dcab30 100644
--- i/src/test/validation_block_tests.cpp
+++ w/src/test/validation_block_tests.cpp
@@ -211,12 +211,66 @@ BOOST_AUTO_TEST_CASE(processnewblock_signals_ordering)
     UnregisterValidationInterface(&sub);
 
     LOCK(cs_main);
     BOOST_CHECK_EQUAL(sub.m_expected_tip, ::ChainActive().Tip()->GetBlockHash());
 }
 
+BOOST_AUTO_TEST_CASE(download_blocks_out_of_order)
+{
+    const int numberOfExtraBlocks = 10;
+    std::vector<std::shared_ptr<const CBlock>> extraBlocks;
+
+    BuildChain(Params().GenesisBlock().GetHash(), numberOfExtraBlocks, 0, 0, numberOfExtraBlocks, extraBlocks);
+
+    assert(extraBlocks.size() == numberOfExtraBlocks);
+
+    bool ignored;
+    BlockValidationState state;
+    std::vector<CBlockHeader> headers;
+    std::transform(extraBlocks.begin(), extraBlocks.end(), std::back_inserter(headers), [](std::shared_ptr<const CBlock> b) { return b->GetBlockHeader(); });
+
+    // Process all the headers so we understand the toplogy of the chain
+    BOOST_CHECK(ProcessNewBlockHeaders(headers, state, Params()));
+
+    // Connect the genesis block and drain any outstanding events
+    BOOST_CHECK(ProcessNewBlock(Params(), std::make_shared<CBlock>(Params().GenesisBlock()), true, &ignored));
+    SyncWithValidationInterfaceQueue();
+
+    // subscribe to events (this subscriber will validate event ordering)
+    const CBlockIndex* initial_tip = nullptr;
+    {
+        LOCK(cs_main);
+        initial_tip = ::ChainActive().Tip();
+    }
+    TestSubscriber sub(initial_tip->GetBlockHash());
+    RegisterValidationInterface(&sub);
+
+    // Just 4 blocks fit in one blk file
+
+    //std::vector<size_t> order{0, 1, 2, 3,  4, 5, 6, 7,  8, 9, 10};
+    std::vector<size_t> order{0, 1, 2, 3,  4, 5, 6, 8,  7, 9, 10};
+    //std::vector<size_t> order{0, 1, 2, 3,  4, 5, 6, 7,  9, 8, 10};
+    //std::vector<size_t> order{0, 1, 2, 3,  4, 7, 8, 9,  10, 5, 6};
+    //std::vector<size_t> order{0, 2, 3, 4,  5, 6, 7, 8,  9, 1, 10};
+
+    for (size_t i = 1; i <= numberOfExtraBlocks; ++i) {
+        // Block at height H is in extraBlocks[H - 1].
+        bool ok = ProcessNewBlock(Params(), extraBlocks[order[i] - 1], true, &ignored);
+        assert(ok);
+    }
+
+    while (GetMainSignals().CallbacksPending() > 0) {
+        MilliSleep(100);
+    }
+
+    UnregisterValidationInterface(&sub);
+
+    LOCK(cs_main);
+    BOOST_CHECK_EQUAL(sub.m_expected_tip, ::ChainActive().Tip()->GetBlockHash());
+}
+
 /**
  * Test that mempool updates happen atomically with reorgs.
  *
  * This prevents RPC clients, among others, from retrieving immediately-out-of-date mempool data
  * during large reorgs.
  *
diff --git i/src/validation.cpp w/src/validation.cpp
index faf92b3b0..0ff0dbcd9 100644
--- i/src/validation.cpp
+++ w/src/validation.cpp
@@ -1731,40 +1731,44 @@ DisconnectResult CChainState::DisconnectBlock(const CBlock& block, const CBlockI
     return fClean ? DISCONNECT_OK : DISCONNECT_UNCLEAN;
 }
 
 static void FlushUndoFile(int block_file, bool finalize = false)
 {
     FlatFilePos undo_pos_old(block_file, vinfoBlockFile[block_file].nUndoSize);
+    LogPrintf("BBB FlushUndoFile(): rev%05u pos=%u, finalize=%d\n", undo_pos_old.nFile, undo_pos_old.nPos, finalize);
     if (!UndoFileSeq().Flush(undo_pos_old, finalize)) {
         AbortNode("Flushing undo file to disk failed. This is likely the result of an I/O error.");
     }
 }
 
 static void FlushBlockFile(bool fFinalize = false, bool finalize_undo = false)
 {
     LOCK(cs_LastBlockFile);
     FlatFilePos block_pos_old(nLastBlockFile, vinfoBlockFile[nLastBlockFile].nSize);
+    LogPrintf("BBB FlushBlockFile(): blk%05u pos=%u, finalize=%d\n", block_pos_old.nFile, block_pos_old.nPos, fFinalize);
     if (!BlockFileSeq().Flush(block_pos_old, fFinalize)) {
         AbortNode("Flushing block file to disk failed. This is likely the result of an I/O error.");
     }
     // we do not always flush the undo file, as the chain tip may be lagging behind the incoming blocks,
     // e.g. during IBD or a sync after a node going offline
     if (!fFinalize || finalize_undo) FlushUndoFile(nLastBlockFile, finalize_undo);
 }
 
-static bool FindUndoPos(BlockValidationState &state, int nFile, FlatFilePos &pos, unsigned int nAddSize);
+static bool FindUndoPos(BlockValidationState &state, int nFile, FlatFilePos &pos, unsigned int nAddSize, int nHeight);
 
 static bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationState& state, CBlockIndex* pindex, const CChainParams& chainparams)
 {
     // Write undo information to disk
     if (pindex->GetUndoPos().IsNull()) {
         FlatFilePos _pos;
-        if (!FindUndoPos(state, pindex->nFile, _pos, ::GetSerializeSize(blockundo, CLIENT_VERSION) + 40))
+        if (!FindUndoPos(state, pindex->nFile, _pos, ::GetSerializeSize(blockundo, CLIENT_VERSION) + 40, pindex->nHeight))
             return error("ConnectBlock(): FindUndoPos failed");
         if (!UndoWriteToDisk(blockundo, _pos, pindex->pprev->GetBlockHash(), chainparams.MessageStart()))
             return AbortNode(state, "Failed to write undo data");
+        LogPrintf("BBB WriteUndoDataForBlock(): _pos.nFile=%d, nLastBlockFile=%d, pindex->height=%d, last height in file=%d\n",
+            _pos.nFile, nLastBlockFile, pindex->nHeight, vinfoBlockFile[_pos.nFile].nHeightLast);
         // rev files are written in block height order, whereas blk files are written as blocks come in (often out of order)
         // we want to flush the rev (undo) file once we've written the last block, which is indicated by the last height
         // in the block file info as below; note that this does not catch the case where the undo writes are keeping up
         // with the block writes (usually when a synced up node is getting newly mined blocks) -- this case is caught in
         // the FindBlockPos function
         if (_pos.nFile < nLastBlockFile && static_cast<uint32_t>(pindex->nHeight) == vinfoBlockFile[_pos.nFile].nHeightLast) {
@@ -3246,12 +3250,15 @@ static bool FindBlockPos(FlatFilePos &pos, unsigned int nAddSize, unsigned int n
             }
         }
         pos.nFile = nFile;
         pos.nPos = vinfoBlockFile[nFile].nSize;
     }
 
+    LogPrintf("BBB FindBlockPos(): height=%u, belongs to blk%05u[%u..%u]\n",
+        nHeight, pos.nFile, pos.nPos, pos.nPos + nAddSize - 1);
+
     if ((int)nFile != nLastBlockFile) {
         if (!fKnown) {
             LogPrintf("Leaving block file %i: %s\n", nLastBlockFile, vinfoBlockFile[nLastBlockFile].ToString());
         }
         FlushBlockFile(!fKnown, finalize_undo);
         nLastBlockFile = nFile;
@@ -3275,22 +3282,25 @@ static bool FindBlockPos(FlatFilePos &pos, unsigned int nAddSize, unsigned int n
     }
 
     setDirtyFileInfo.insert(nFile);
     return true;
 }
 
-static bool FindUndoPos(BlockValidationState &state, int nFile, FlatFilePos &pos, unsigned int nAddSize)
+static bool FindUndoPos(BlockValidationState &state, int nFile, FlatFilePos &pos, unsigned int nAddSize, int nHeight)
 {
     pos.nFile = nFile;
 
     LOCK(cs_LastBlockFile);
 
     pos.nPos = vinfoBlockFile[nFile].nUndoSize;
     vinfoBlockFile[nFile].nUndoSize += nAddSize;
     setDirtyFileInfo.insert(nFile);
 
+    LogPrintf("BBB FindUndoPos(): height=%d, belongs to: rev%05u[%u..%u]\n",
+        nHeight, nFile, pos.nPos, pos.nPos + nAddSize - 1);
+
     bool out_of_space;
     size_t bytes_allocated = UndoFileSeq().Allocate(pos, nAddSize, out_of_space);
     if (out_of_space) {
         return AbortNode(state, "Disk space is too low!", _("Error: Disk space is too low!").translated, CClientUIInterface::MSG_NOPREFIX);
     }
     if (bytes_allocated != 0 && fPruneMode) {
@@ -3703,12 +3713,13 @@ bool ProcessNewBlockHeaders(const std::vector<CBlockHeader>& headers, BlockValid
     }
     return true;
 }
 
 /** Store block on disk. If dbp is non-nullptr, the file is known to already reside on disk */
 static FlatFilePos SaveBlockToDisk(const CBlock& block, int nHeight, const CChainParams& chainparams, const FlatFilePos* dbp) {
+    LogPrintf("BBB SaveBlockToDisk(): height=%d\n", nHeight);
     unsigned int nBlockSize = ::GetSerializeSize(block, CLIENT_VERSION);
     FlatFilePos blockPos;
     if (dbp != nullptr)
         blockPos = *dbp;
     if (!FindBlockPos(blockPos, nBlockSize+8, nHeight, block.GetBlockTime(), dbp != nullptr)) {
         error("%s: FindBlockPos failed", __func__);
diff --git i/src/validation.h w/src/validation.h
index a5335edc4..e45d29ef5 100644
--- i/src/validation.h
+++ w/src/validation.h
@@ -67,17 +67,25 @@ static const unsigned int DEFAULT_DESCENDANT_SIZE_LIMIT = 101;
 static const unsigned int EXTRA_DESCENDANT_TX_SIZE_LIMIT = 10000;
 /** Default for -mempoolexpiry, expiration time for mempool transactions in hours */
 static const unsigned int DEFAULT_MEMPOOL_EXPIRY = 336;
 /** Maximum kilobytes for transactions to store for processing during reorg */
 static const unsigned int MAX_DISCONNECTED_TX_POOL_SIZE = 20000;
 /** The maximum size of a blk?????.dat file (since 0.8) */
-static const unsigned int MAX_BLOCKFILE_SIZE = 0x8000000; // 128 MiB
+// Size chosen so that just 4 blocks from the download_blocks_out_of_order unit
+// tests fit in one blk file.
+static const unsigned int MAX_BLOCKFILE_SIZE = 1200;
 /** The pre-allocation chunk size for blk?????.dat files (since 0.8) */
-static const unsigned int BLOCKFILE_CHUNK_SIZE = 0x1000000; // 16 MiB
+static const unsigned int BLOCKFILE_CHUNK_SIZE = 600;
 /** The pre-allocation chunk size for rev?????.dat files (since 0.8) */
-static const unsigned int UNDOFILE_CHUNK_SIZE = 0x100000; // 1 MiB
+// Size chosen so that one chunk can accommodate 3 undos from the
+// download_blocks_out_of_order unit test. Given that we will have max
+// 4 undos per one undo file it means that when writing the first undo
+// we allocate space for 3 undos, when writing the 4th undo we allocate
+// space for 3 more (6 in total). During finalizing we should shrink
+// the file size from 6 to 4 undos.
+static const unsigned int UNDOFILE_CHUNK_SIZE = 130;
 
 /** Maximum number of dedicated script-checking threads allowed */
 static const int MAX_SCRIPTCHECK_THREADS = 15;
 /** -par default (number of script-checking threads, 0 = auto) */
 static const int DEFAULT_SCRIPTCHECK_THREADS = 0;
 /** Number of blocks that can be requested at any given time from a single peer. */
```

</details>

kallewoof commented:
@vasild Thanks a lot for testing and for helping get this into shape!

I've squashed the commits into one, and also updated the commit description. I am going to post updated results from the node running this in a bit.

vasild commented:
ACK ac94141af (no changes in the code since ed34e00da).

kallewoof commented:
Updated https://github.com/bitcoin/bitcoin/pull/17994#issuecomment-601195517 with more logs. They're not very exciting, but they indicate things are working as they should.

fjahr commented:
Code review re-ACK ac94141af0c16161afa68de1c3720f254ae4e12c

Only changes since my last review is the introduction of the `finalize_undo` parameter and another comments change.

jonatack commented on src/validation.cpp:
```diff
@@ -1731,19 +1731,24 @@ DisconnectResult CChainState::DisconnectBlock(const CBlock& block, const CBlockI
     }
+    // we do not always flush the undo file, as the chain tip may be lagging behind the incoming blocks,
+    // e.g. during IBD or a sync after a node going offline
+    if (!fFinalize || finalize_undo) FlushUndoFile(nLastBlockFile, finalize_undo);
```
`finalize_undo` is clear but it might be good to comment on the `fFinalize` meaning and logic

jonatack commented on src/validation.cpp:
```diff
@@ -3220,8 +3233,13 @@ static bool FindBlockPos(FlatFilePos &pos, unsigned int nAddSize, unsigned int n
         while (vinfoBlockFile[nFile].nSize + nAddSize >= MAX_BLOCKFILE_SIZE) {
+            // when the undo file is keeping up with the block file, we want to flush it explicitly
+            // when it is lagging behind (more blocks arrive than are being connected), we let the
+            // undo block write case handle it
```
it took a couple of readings for me to parse this as paragraph as two sentences rather than one; ISTM adding the missing punctuation would be more clear

jonatack reviewed (commented):
Code review ACK ac94141af0c16 

Also built, ran tests and bitcoind as a minimal sanity check. 

Thanks for fixing this and adding documentation. I think additional/better documentation would be helpful, including but not limited to the small suggestions below and the `while` mentioned above. Happy to re-ACK for that.

Kudos @vasild for the unit test. The code wasn't designed for the components to be easily tested separately, but behavior specs / regression tests, if feasible with some refactoring, would be continuously valuable.

Don't hesitate to ping me for review if you open a follow-up PR to improve the documentation or tests.

EDIT: a further [follow-up idea](https://bitcoincore.reviews/17994.html#l-201) by @sipa from today's Review Club session: benchmark how much the pre-allocation actually helps nowadays (to maybe remove it entirely).

kallewoof commented:
Will address @jonatack points if I need to rebase or modify in other ways.
